File: swift/anagrams.swift
import Foundation

func generatePermutations(of a: inout [Character], upTo n: Int) {
    if n == 0 {
        print(String(a))
    } else {
        for i in 0..<n {
            generatePermutations(of: &a, upTo: n-1)
            a.swapAt(n % 2 == 0 ? 0 : i, n)
        }
        generatePermutations(of: &a, upTo: n-1)
    }
}

if CommandLine.arguments.count != 2 {
    fputs("Exactly one argument is required\n", __stderrp)
    exit(1)
}
let word = CommandLine.arguments[1]
var charArray = Array(word)
generatePermutations(of: &charArray, upTo: charArray.count-1)

File: cpp/anagrams.cpp
#include <iostream>
#include <utility>
#include <cstring>

void generatePermutations(char* s, int n) {
  if (n == 0) {
    std::cout << s << '\n';
  } else {
    for (int i = 0; i < n; i++) {
      generatePermutations(s, n-1);
      int j = n % 2 == 0 ? 0 : i;
      std::swap(s[j], s[n]);
    }
    generatePermutations(s, n-1);
  }
}

int main(int argc, char** argv) {
  if (argc != 2) {
    std::cerr << "Exactly one argument is required\n";
    return 1;
  }
  generatePermutations(argv[1], strlen(argv[1])-1);
}
